### ISSUE TITLE
Improve Start-Process functionality to control the position and size of the new application window

### ISSUE THREAD

**Issue by schicout** [PowerShell/PowerShell#21499]
### The task

I work on several computers with different configurations and different types of monitors, but with a similar set of software. I use Windows 10 with Virtual Desktops (VD). There are three VD. First VD - Visual Studio, SSMS, Windows Terminal. The second VD used for mail, MS Edge and an excel file, the third VD used for MS Teams, another window of Edge, another excel file, telegram, whatsapp. And all this is laid out in a certain way on the VDs, because a 4k monitor allows you to organize it all very well. When I turn off the computer, I use hibernation mode, so when I turn it on, this whole order is well preserved. But when an Windows update "arrives" that requires rebooting the computer, I have to play this Tetris, starting and arranging applications on VDs, because not all applications can immediately be located where they were last time, and even if they can, they confuse the VD e.t.c...

### Possible decisions

I found a script that allows me to control the position and size of the application window that is just launched, but does not allow me to specify the VD on which I would like the application to appear. In addition, it requires some work for this script to work with both "usial-way-installed" applications and applications received from the MS Store.

```
Add-Type @"
  using System;
  using System.Runtime.InteropServices;
  public class Win32 {
    [DllImport("user32.dll")]
    [return: MarshalAs(UnmanagedType.Bool)]
    public static extern bool GetWindowRect(IntPtr hWnd, out RECT lpRect);
    [DllImport("user32.dll")]
    [return: MarshalAs(UnmanagedType.Bool)]
    public static extern bool MoveWindow(IntPtr hWnd, int X, int Y, int nWidth, int nHeight, bool bRepaint);
    [DllImport("user32.dll")]
    public static extern IntPtr FindWindow(string ClassName, IntPtr  TitleApp);
  }
  public struct RECT
  {
    public int Left;        // x position of upper-left corner
    public int Top;         // y position of upper-left corner
    public int Right;       // x position of lower-right corner
    public int Bottom;      // y position of lower-right corner
  }
"@
 
start notepad
sleep -Milliseconds 50
$h =[Win32]::FindWindow("Notepad",[IntPtr]::Zero)
#$h = (Get-Process | where {$_.MainWindowTitle -match "Блокнот"}).MainWindowHandle
 
if ($h -eq [IntPtr]::Zero) {return "Window not found"}
$rcWindow = New-Object RECT
[void][Win32]::GetWindowRect($h,[ref]$rcWindow)
$left = 0
$top  = 0
$WndWidth  = $rcWindow.Right - $rcWindow.Left;
$WndHeight = $rcWindow.Bottom -$rcWindow.Top;
[void][Win32]::MoveWindow($h, $left, $top, $WndWidth, $WndHeight, $true)

```

### Feature Request 

Now Posershell can handle "-WindowStyle" parameter of "Start-Process" comandlet. Why not go further and add parameters that would point to the exact starting position of the window, its dimensions and the index (name) of the VD on which this app must be placed. The form (i mean C# class Form) has a StartPosition property with the default value of WindowsDefaultLocation, which means Windows can easily manage these parameters, and if OS can manage them, why not allow the user running the application to manage them too? For example:

`Start-Process Notepad -VirtualDesktopId 2 -InitLocation 100 250 -InitSize 1024 768`

It seems to me that such a solution will be much more elegant and convenient than the hundredth invention of an unreliable bicycle.

**Comment by rhubarb-geek-nz** [comment]
**TL/DR**

PowerShell is a C# cross-platform console application. [StartProcessCommand](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs) uses [ProcessStartInfo](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo?view=net-8.0) which does not include any coordinates.

**Long answer**.

PowerShell, `pwsh.exe` is a console application. What that means is it does not provide any GUI, the OS must provide that through some form of Console or Terminal window, and it communicates with the user by reading from stdin and writing to `stdout` and `stderr`.

It is C# application and uses .NET 8.0. The API to start a process in .NET is [Process.Start](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.start?view=net-8.0) with [ProcessStartInfo](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo?view=net-8.0). This does not contain any graphical location information.

It does however contain [WindowStyle](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.windowstyle?view=net-8.0#system-diagnostics-processstartinfo-windowstyle).

PowerShell already exposes those options.

```
PS> Get-Command Start-Process -Syntax

Start-Process [-FilePath] <string> [[-ArgumentList] <string[]>] [-Credential <pscredential>] [-WorkingDirectory <string>] [-LoadUserProfile] [-NoNewWindow] [-PassThru] [-RedirectStandardError <string>] [-RedirectStandardInput <string>] [-RedirectStandardOutput <string>] [-WindowStyle <ProcessWindowStyle>] [-Wait] [-UseNewEnvironment] [-Environment <hashtable>] [-WhatIf] [-Confirm] [<CommonParameters>]

Start-Process [-FilePath] <string> [[-ArgumentList] <string[]>] [-WorkingDirectory <string>] [-PassThru] [-Verb <string>] [-WindowStyle <ProcessWindowStyle>] [-Wait] [-Environment <hashtable>] [-WhatIf] [-Confirm] [<CommonParameters>]
```

Now you do have the option of more control with [wt.exe](https://apps.microsoft.com/detail/9n0dx20hk701?rtc=1&hl=en-gb&gl=GB), the Windows Terminal is a GUI application. Type wt.exe --help and you get a dialog box, not a print out in the console window. With that you can control location of terminal windows.

> The form (i mean C# class Form) has a StartPosition property

Now yes, in C# you can create Forms for a GUI program. But PowerShell is not one of those. You can write GUIs in C# in Silverlight, WinForms, WPF, XAML, UMP, Maui, WinUI3 or whatever and are welcome to.

Now if you were writing using C or C++ you could use [CreateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw) with [STARTUPINFO](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/ns-processthreadsapi-startupinfow) and have more options to set.

However there is no guarantee that any application will actually abide by the coordinate recommendations in STARTUPINFO.

So you are welcome to write a PowerShell cmdlet that uses DllImport etc to use CreateProcessW with STARTUPINFOW and see how far that takes you, and publish in PSGallery if you think it works well.

There is no need for it to be a core PowerShell component because it is not needed by most people, it is not cross platform, it is not what console programs need and it would always be wrong or insufficient for your needs. It is an ideal feature to add as a PowerShell module from the PSGallery.

> It seems to me that such a solution will be much more elegant and convenient than the hundredth invention of an unreliable bicycle.

Don't ask console application developers to write GUI code, it makes everyone sad and disappointed. If you know what you want, write it (eg Start-ProcessWithStartInfo and Set-WindowPosition) and put it in PSGallery, at least you will be happy.

**Comment by schicout** [comment]
Thank you very much for wasting your time on me and giving such a detailed answer, you have slightly reduced the chaos in my mind about PS and WT. I may have cleared up my misconceptions. At least you gave a lot more useful information than I could Google. Well, or I don’t know how to Google :) Anyway, thank you very much again and don’t be so upset over nothing - doctors say that nerve cells are not restored. 
Have a great day!
Kind regards.
